### PR TITLE
Fix double-scanning behavior for SVGA/VESA games

### LINKED ISSUE #3416: Wrong scaling in Wing Commander 3
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

_No response_

### What Operating System are you using?

Windows 10

### If Other OS, please describe

_No response_

### Relevant hardware info

_No response_

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Wing Commander 3 is only showing part of its display area in SVGA:

image0001

Pressing the Win key fixes it:

image0002

Proceeding to the flight control breaks it again:

image0003

This time alt-tabbing or pressing Win has no effect.

Introduction video is also cut in half and displayed in the bottom part of the screen:

image0004

UNIVBE has no effect on this issue.

Console insists the original display is 640x240 256 colors and gets scaled to 1920x1440 pixels with 1:2 (1:2) pixel aspect ratio (instead of 640x480 pixels).

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Install Wing Commander 3 with SVGA options enabled. 
2. Run with WC3.EXE as per usual.

### Download URL of affected game or software

_No response_

### Your configuration

```ini
machine            = svga_s3
aspect             = auto
integer_scaling    = vertical
viewport           = fit
```

### Provide a Log

_No response_

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

### PR DESCRIPTION
**Pull request by weirddan455**
# Description

The branch which sometimes skips VGA re-sizing for double-scan modes does not work at SVGA/VESA resolutions.  Those modes have a "faked" double-scan that gets calculated here:

https://github.com/dosbox-staging/dosbox-staging/blob/0c8e8cec90cdbda8821492e1b3094b58d11746cd/src/hardware/vga_draw.cpp#L2108-L2119

Simply doubling `vga.draw.address_line_total` is not enough.

## Related issues

Fixes #3416

# Manual testing

Confirmed this fixes Wing Commander 3.  That game switches into double-scanning (640X240) while playing videos and then back out again (640X480) when the video finishes playing.  These transitions were not being handled properly before.

Regression Testing:

- Watched **Second Reality** demo.  Looks fine.
- **Catacomb 3D** works
- **Legend of Kyrandia** scrolling bug is not present
- **Threat** works

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/hardware/vga_crtc.cpp [hunk 1]
```diff
@@ -26,6 +26,12 @@
 #include "pic.h"
 #include "vga.h"
 
+static bool is_svga_or_vesa(VGAModes mode)
+{
+	return mode == M_LIN4 || mode == M_LIN8 || mode == M_LIN15 ||
+		mode == M_LIN16 || mode == M_LIN24 || mode == M_LIN32;
+}
+
```
Comment (weirddan455) [inline src/hardware/vga_crtc.cpp:34]: The C hacker in me kind of wants to change this to:

```c
return mode & (M_LIN4 | MLIN_8 | M_LIN15 | M_LIN16 | M_LIN24 | M_LIN32);
```
Since this enum is also a bitfield.  At the same, mis-reading what bitwise instructions are doing has been a problem.

File: src/hardware/vga_crtc.cpp [hunk 2]
```diff
@@ -189,21 +195,21 @@ void vga_write_p3d5(io_port_t, io_val_t value, io_width_t)
...
-			// Only true if only `start_vertical_blanking_bit9` has
-			// been changed and all other bits are unchanged.
-			if (old_val.data == flipped.data) {
-				VGA_StartResize();
-			}
-
+		if (vga.draw.is_double_scanning && !is_svga_or_vesa(vga.mode)) {
+			// This is a potential fast path where we can skip VGA resizing.
+			// A rare few games (ex. Wing Commander 3) fiddle with the double-scan bit
+			// in SVGA/VESA modes.
```
Comment (johnnovak) [inline src/hardware/vga_crtc.cpp:201]: I'd prefer if you forced a resizing as I don't like it if this algorithm becomes too "distributed". We don't want a "fast" path here as the resize logic is fast enough, but we absolutely want as much clarity as we can achieve.

What should happen here is detecting the conditions that require a resize, then modify the big resize function (if needed; maybe triggering a resize here is what's missing).

Comment (weirddan455) [reply src/hardware/vga_crtc.cpp:201]: "Fast path" is probably the wrong wording as it's not really about performance.  My first thought was just to call `VGA_StartResize` all the time but you tried that and it apparently broke Legend of Kyriandia then #3279 restored this behavior.

I guess I'm unclear what you mean if not that.

File: src/hardware/vga_crtc.cpp (hunk 1 above)
Comment (weirddan455) [reply src/hardware/vga_crtc.cpp:34]: I figured you would say something like that :smile:

File: src/hardware/vga_crtc.cpp (hunk 2 above)
Comment (johnnovak) [reply src/hardware/vga_crtc.cpp:201]: Okay, just proceed as you think is best then we'll talk at review time 😄

Comment (weirddan455) [reply src/hardware/vga_crtc.cpp]: I think I see what you mean now.  I did some regression testing and I think we can get rid of this special double-scan path.

If you avoid all checking, the Legend of Kyrandia bug returns.  However, if you keep only the bottom (general) path, everything seems to work.

I force pushed this to a single commit to remove the top branch.  It is nice when you can fix a bug simply by deleting code :smile:

Comment (johnnovak) [reply src/hardware/vga_crtc.cpp]: Yeah, this looks good. I meant only to trigger the resize from here, then updating the VGA state based on the register value change should be done in the big `setup_drawing()` function in `vga_draw.cpp` (see below).

If this simplification works, then great 🎉 I will need to read `setup_drawing()` again in light of this change for correctness, then test with a bunch of special VGA programs that do weird things to the hardware... The good news is the outcome of that will be a VGA regression testing wiki page that we sorely need.

Great job so far @weirddan455 👍🏻 Good to see I'm not the only one hacking away at the VGA code.

```cpp
// Can change the following VGA state:
//
//   vga.draw.address_line_total
//   vga.draw.blocks
//   vga.draw.delay.hdend
//   vga.draw.dos_refresh_hz
//   vga.draw.line_length
//   vga.draw.linear_base
//   vga.draw.linear_mask
//   vga.draw.mode
//   vga.draw.resizing
//   vga.draw.uses_vga_palette
//   vga.draw.vblank_skip
//   vga.draw.vret_triggered
//
ImageInfo setup_drawing()
{
```

Comment (weirddan455) [reply src/hardware/vga_crtc.cpp]: `vga.draw.address_line_total` (only VGA variable that previously got set here) already gets re-set in `setup_drawing()` according to the register values (and the code there correctly handles edge cases unlike here).  I don't think anything should need changing there.

It seems this "fast path" logic existed before your improvements to double-scanning and so is incorrect now.  Hopefully there won't be any problems just removing it as that would be the simplest solution.

Comment (johnnovak) [reply src/hardware/vga_crtc.cpp]: > It seems this "fast path" logic existed before your improvements to double-scanning and so is incorrect now.

Yeah, and I didn't like it, just was reluctant to touch it 😄 

> vga.draw.address_line_total (only VGA variable that previously got set here) already gets re-set in setup_drawing() according to the register values (and the code there correctly handles edge cases unlike here). I don't think anything should need changing there.

That's my hope. I'll just need to test the crap out of it 😄

**Comment by johnnovak** [comment]
Okay, I've expanded the video emulation tests page with lots of VGA tests based on our recent experience with this thing:
https://github.com/dosbox-staging/dosbox-staging/wiki/Video-emulation-tests#vga

I've done pretty much all of that, specifically:

- Testing all 15/16-bit high colour and 24/32-bit true colour VESA modes in QPV.
- All VGA modes in Deluxe Paint II
- Most of the games I added to the VGA tests section.

**Review by johnnovak — APPROVED** [review]
Nice one @weirddan455

**Comment by weirddan455** [comment]
Thanks for the tests and the documentation @johnnovak Hopefully this is the last time we have to touch this function for a while.
